### PR TITLE
FFM-12192 Major version revision to support Ruby 3

### PR DESCRIPTION
**Pull request by erdirowlands**
# What - breaking changes
This introduces breaking changes to the SDK, specifically:

- All public methods now enforce keyword arguments
- Critical path code uses keyword arguments
- SDK enforces Singleton pattern

Why?

- Keywords: Ruby 3 changes how keyword and positional arguments are used together.  We are standardising on keyword args for V2 for public API methods and internal critical path.
- Singleton: Our SDKs don't officially support multi-instance, with Erlang being an exception, so this enforces that at the SDK level. 

# What - other changes

- When cloning and clearing the evaluations metrics map, removes per key locking in favour of simplified clear and clone, to align with target metrics maps. 
- Add validation for evaluation metrics processing
- Add validation for comparing two metrics events

- TIdy up default variation behaviour and logs for
    -  Flag not being found
    - Wait for init timing out
  
# Testing
- Unit tests updated 
- Manual testing
- Testgrid integration suite

### CODE REVIEW

File: example/simple_example/example.rb [hunk 1]
```diff
@@ -1,101 +0,0 @@
-require "logger"
```
Comment (erdirowlands) [inline example/simple_example/example.rb:1]: Deleted on purpose.

File: lib/ff/ruby/server/sdk/api/cf_client.rb [hunk 2]
```diff
@@ -1,91 +1,63 @@
...
-
-      @config = config
+  include Singleton
+
+  def init(api_key: nil, config: nil, connector: nil)
+    # Only initialize if @client is nil to avoid reinitialization
+    unless @client
+      @config = config || ConfigBuilder.new.build
+      @client = InnerClient.new(api_key, @config, connector)
+      @config.logger.debug "Client initialized with API key: #{api_key}"
```
Comment (phoffer) [inline lib/ff/ruby/server/sdk/api/cf_client.rb:14]: I think it would be convenient to always return `@client` from `#init`

Comment (erdirowlands) [reply lib/ff/ruby/server/sdk/api/cf_client.rb:14]: Yes, great spot @phoffer .  Thanks 👍

**Comment by erdirowlands** [comment]
Closing as now outdated.  Will re-open later if required.

**Pull request closed without merging**